File: src/parsing/functions/max.ts
import AggregateFunction from "./aggregate_function";
import { FunctionDef } from "./function_metadata";
import ReducerElement from "./reducer_element";

class MaxReducerElement extends ReducerElement {
    private _value: any = null;
    public get value(): any {
        return this._value;
    }
    public set value(value: any) {
        if (this._value === null || value > this._value) {
            this._value = value;
        }
    }
}

@FunctionDef({
    description: "Returns the maximum value across grouped rows",
    category: "aggregate",
    parameters: [{ name: "value", description: "Value to compare", type: "number" }],
    output: { description: "Maximum value", type: "number", example: 10 },
    examples: ["WITH [3, 1, 2] AS nums UNWIND nums AS n RETURN max(n)"],
})
class Max extends AggregateFunction {
    constructor() {
        super("max");
        this._expectedParameterCount = 1;
    }
    public reduce(element: MaxReducerElement): void {
        element.value = this.firstChild().value();
    }
    public element(): MaxReducerElement {
        return new MaxReducerElement();
    }
}

export default Max;

File: flowquery-py/src/parsing/functions/max_.py
"""Max aggregate function."""

from typing import Any

from .aggregate_function import AggregateFunction
from .function_metadata import FunctionDef
from .reducer_element import ReducerElement


class MaxReducerElement(ReducerElement):
    """Reducer element for Max aggregate function."""

    def __init__(self) -> None:
        self._value: Any = None

    @property
    def value(self) -> Any:
        return self._value

    @value.setter
    def value(self, val: Any) -> None:
        if self._value is None or val > self._value:
            self._value = val


@FunctionDef({
    "description": "Returns the maximum value across grouped rows",
    "category": "aggregate",
    "parameters": [
        {"name": "value", "description": "Value to compare", "type": "number"}
    ],
    "output": {"description": "Maximum value", "type": "number", "example": 10},
    "examples": ["WITH [3, 1, 2] AS nums UNWIND nums AS n RETURN max(n)"]
})
class Max(AggregateFunction):
    """Max aggregate function.

    Returns the maximum value across grouped rows.
    """

    def __init__(self) -> None:
        super().__init__("max")
        self._expected_parameter_count = 1

    def reduce(self, element: MaxReducerElement) -> None:
        element.value = self.first_child().value()

    def element(self) -> MaxReducerElement:
        return MaxReducerElement()
